# Natural Language Processing Project

# Multi-Label Categorical Classification of English Quotes

<b>Rish (Rishabh) Kaushick</b> <br>
College of Engineering <br>
Northeastern University <br>
Toronto, ON <br>
kaushick.r@northeastern.edu

# Downloading The Dataset From Hugging Face

In [1]:
from huggingface_hub import hf_hub_download
import pandas as pd

REPO_ID = "Abirate/english_quotes"
FILENAME = "quotes.jsonl"

# since the data is in jsonl format, we read it using read_json() function with lines=True
dataset = pd.read_json(hf_hub_download(repo_id=REPO_ID, filename=FILENAME, repo_type="dataset"), lines=True)


In [2]:
dataset

quote  \
0        “Be yourself; everyone else is already taken.”   
1     “I'm selfish, impatient and a little insecure....   
2     “Two things are infinite: the universe and hum...   
3                      “So many books, so little time.”   
4     “A room without books is like a body without a...   
...                                                 ...   
2503  “Morality is simply the attitude we adopt towa...   
2504  “Don't aim at success. The more you aim at it ...   
2505  “In life, finding a voice is speaking and livi...   
2506  “Winter is the time for comfort, for good food...   
2507                      “Silence is so freaking loud”   

                     author                                               tags  
0               Oscar Wilde  [be-yourself, gilbert-perreira, honesty, inspi...  
1            Marilyn Monroe  [best, life, love, mistakes, out-of-control, t...  
2           Albert Einstein  [human-nature, humor, infinity, philosophy, sc...  
3               Frank Zappa                                     [books, humor]  
4     Marcus Tullius Cicero                              [books, simile, soul]  
...                     ...                                                ...  
2503           Oscar Wilde,                             [morality, philosophy]  
2504      Viktor E. Frankl,                               [happiness, success]  
2505           John Grisham                               [inspirational-life]  
2506          Edith Sitwell                            [comfort, home, winter]  
2507          Sarah Dessen,    [just-listen, loud, owen, sara-dessen, silence]  

[2508 rows x 3 columns]

# Exploratory Data Analysis

In [3]:
dataset.head(10)

quote                 author  \
0     “Be yourself; everyone else is already taken.”            Oscar Wilde   
1  “I'm selfish, impatient and a little insecure....         Marilyn Monroe   
2  “Two things are infinite: the universe and hum...        Albert Einstein   
3                   “So many books, so little time.”            Frank Zappa   
4  “A room without books is like a body without a...  Marcus Tullius Cicero   
5  “Be who you are and say what you feel, because...      Bernard M. Baruch   
6  “You've gotta dance like there's nobody watchi...      William W. Purkey   
7  “You know you're in love when you can't fall a...              Dr. Seuss   
8  “You only live once, but if you do it right, o...               Mae West   
9  “Be the change that you wish to see in the wor...         Mahatma Gandhi   

                                                tags  
0  [be-yourself, gilbert-perreira, honesty, inspi...  
1  [best, life, love, mistakes, out-of-control, t...  
2  [human-nature, humor, infinity, philosophy, sc...  
3                                     [books, humor]  
4                              [books, simile, soul]  
5  [ataraxy, be-yourself, confidence, fitting-in,...  
6  [dance, heaven, hurt, inspirational, life, lov...  
7                     [dreams, love, reality, sleep]  
8                                      [humor, life]  
9  [action, change, inspirational, philosophy, wish]

In [4]:
# setting random_state in the random sample, so that each time the same value is presented
dataset.sample(5, random_state=23)

quote             author  \
2062  “Why are you worrying about YOU-KNOW-WHO, when...      J.K. Rowling,   
1078  “I am not sure that I exist, actually. I am al...  Jorge Luis Borges   
1176  “If you're lonely when you're alone, you're in...   Jean-Paul Sartre   
414   “You educate a man; you educate a man. You edu...      Brigham Young   
508   “There is neither happiness nor misery in the ...    Alexandre Dumas   

                                   tags  
2062                            [humor]  
1078                             [life]  
1176             [loneliness, solitude]  
414   [education, feminism, men, women]  
508       [hope, inspirational, wisdom]

In [5]:
dataset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2508 entries, 0 to 2507
Data columns (total 3 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   quote   2508 non-null   object
 1   author  2508 non-null   object
 2   tags    2508 non-null   object
dtypes: object(3)
memory usage: 58.9+ KB


In [6]:
# Quotes by J.k. Rowling
dataset.loc[dataset['author'] == 'J.K. Rowling,']

quote         author  \
11    “If you want to know what a man's like, take a...  J.K. Rowling,   
31    “It is our choices, Harry, that show what we t...  J.K. Rowling,   
35    “It does not do to dwell on dreams and forget ...  J.K. Rowling,   
53          “I solemnly swear that I am up to no good.”  J.K. Rowling,   
68    “It takes a great deal of bravery to stand up ...  J.K. Rowling,   
...                                                 ...            ...   
2321              “Don't let the muggles get you down.”  J.K. Rowling,   
2342  “I'm going to keep going until I succeed — o...  J.K. Rowling,   
2401  “You think the dead we loved ever truly leave ...  J.K. Rowling,   
2472  “Hello, Harry" said George, beaming at him. "W...  J.K. Rowling,   
2478  “But Dumbledore says he doesn't care what they...  J.K. Rowling,   

                                      tags  
11               [character, sirius-black]  
31                    [abilities, choices]  
35                          [dreams, life]  
53                              [mischief]  
68    [albus-dumbledore, courage, friends]  
...                                    ...  
2321         [encouragement, muggles, wit]  
2342                        [harry-potter]  
2401                                [dead]  
2472                              [humour]  
2478  [celebrity, dumbledore, fame, funny]  

[74 rows x 3 columns]

## Exploring The Target Variable - 'tags'

In [7]:
# Understanding the 'tags' attribute:
dataset['tags']

0       [be-yourself, gilbert-perreira, honesty, inspi...
1       [best, life, love, mistakes, out-of-control, t...
2       [human-nature, humor, infinity, philosophy, sc...
3                                          [books, humor]
4                                   [books, simile, soul]
                              ...                        
2503                               [morality, philosophy]
2504                                 [happiness, success]
2505                                 [inspirational-life]
2506                              [comfort, home, winter]
2507      [just-listen, loud, owen, sara-dessen, silence]
Name: tags, Length: 2508, dtype: object

In [8]:
dataset['tags'][0]

['be-yourself',
 'gilbert-perreira',
 'honesty',
 'inspirational',
 'misattributed-oscar-wilde',
 'quote-investigator']

In [9]:
dataset['tags'][0][0]

'be-yourself'

In [10]:
dataset['tags'][0][3]

'inspirational'

From the above we can see that the values in the 'tags' attribute are stored as a list.

In [11]:
# Looking for how many unique values are present in the author column
len(dataset['author'].unique())

880

In [12]:
dataset['tags'].value_counts()

tags
[love]                                                                                                                                                                                                                          75
[inspirational]                                                                                                                                                                                                                 63
[humor]                                                                                                                                                                                                                         37
[life]                                                                                                                                                                                                                          34
[writing]                                                                              

The above does not help us much because one row could have multiple categories in the list. <br>
Therefore, in order to see how many unique tags are there, we might need to iterate through the entire pandas dataframe, and store them in a Python set.

In [13]:
# iterating through the values in the pandas dataframe
tags = dataset['tags']
print(tags)
print(type(tags))

0       [be-yourself, gilbert-perreira, honesty, inspi...
1       [best, life, love, mistakes, out-of-control, t...
2       [human-nature, humor, infinity, philosophy, sc...
3                                          [books, humor]
4                                   [books, simile, soul]
                              ...                        
2503                               [morality, philosophy]
2504                                 [happiness, success]
2505                                 [inspirational-life]
2506                              [comfort, home, winter]
2507      [just-listen, loud, owen, sara-dessen, silence]
Name: tags, Length: 2508, dtype: object
<class 'pandas.core.series.Series'>


In [14]:
# Iterting though the pandas Series object and adding each value in a Set
# The benefit of the set is that only the unique values will be added, therefore all duplicates will be removed
unique_tags = set()

# row_tags is the list of tags in each row
# we can see that sometimes a single row has multiple tags, therefore, we will nest another for loop inside which reads individual tags
for row_tags in tags:
    for tag in row_tags:
        unique_tags.add(tag)

print(unique_tags)

{'inej-ghafa', 'living-now', 'jordan-kyle', 'immortals', 'trial', 'chaol-westfall', 'role-models', 'misattributed-to-albert-camus', 'frienship', 'knowledge-power', 'zoe-nightshade', 'the-lightning-thief', 'perspective', 'simple', 'dance', 'distractions', 'weight', 'soul-ties', 'self-esteem', 'clock', 'belief', 'winnie', 'be-yourself-quotes', 'imperfection', 'how-much', 'malfoy', 'mystery', 'responsibility', 'suess', 'selfishness', 'move-on', 'independent-thought', 'true-family', 'hermione', 'stand', 'remembrance', 'house', 'reality', 'introspection', 'promise', 'habits', 'voting', 'paradise', 'cleanliness', 'education', 'dreamers', 'first-sentence', 'reality-check', 'alaska-young', 'hamlet', 'flying-spirit', 'struggle', 'broad-mindedness', 'caring', 'lesson-on-living', 'atheist', 'confidences', 'acceptance', 'governments', 'gift', 'inspirational-stories', 'dictionary', 'claire-fraser', 'wife', 'shampoo', 'lupin', 'pijamas', 'innocence', 'adaptation', 'rosemarie-hathaway', 'sharing', 't

In [15]:
len(unique_tags)

2161

There are a total of 2,161 tags. This is too many if we want this to be our target variable.

In [16]:
# Let me create a Python dictionary to store the tag and it's corresponding frequency throughout this dataset.
tag_freq = {}

for row_tag in tags:
    for tag in row_tag:
        if tag in tag_freq:
            tag_freq[tag] += 1
        else:
            tag_freq[tag] = 1

print(tag_freq)

{'be-yourself': 16, 'gilbert-perreira': 1, 'honesty': 12, 'inspirational': 319, 'misattributed-oscar-wilde': 2, 'quote-investigator': 1, 'best': 1, 'life': 295, 'love': 327, 'mistakes': 10, 'out-of-control': 1, 'truth': 61, 'worst': 1, 'human-nature': 9, 'humor': 254, 'infinity': 4, 'philosophy': 60, 'science': 26, 'stupidity': 7, 'universe': 6, 'books': 141, 'simile': 8, 'soul': 12, 'ataraxy': 4, 'confidence': 11, 'fitting-in': 1, 'individuality': 21, 'misattributed-dr-seuss': 1, 'those-who-matter': 1, 'dance': 5, 'heaven': 2, 'hurt': 8, 'sing': 2, 'dreams': 31, 'reality': 17, 'sleep': 7, 'action': 7, 'change': 34, 'wish': 2, 'character': 11, 'sirius-black': 2, 'friends': 28, 'friendship': 63, 'misattributed-albert-camus': 1, 'lies': 12, 'lying': 6, 'memory': 9, 'friend': 3, 'knowledge': 17, 'enemies': 4, 'forgiveness': 10, 'strategy': 3, 'carpe-diem': 7, 'education': 22, 'learning': 15, 'darkness': 7, 'drive-out': 1, 'hate': 10, 'light': 11, 'peace': 14, 'music': 31, 'intelligence': 

In [17]:
# Let us sort the dictionary in descending order of the frequencies:
tag_freq_sorted = sorted(tag_freq.items(), key=lambda x:x[1], reverse=True)
print(tag_freq_sorted)

[('love', 327), ('inspirational', 319), ('life', 295), ('humor', 254), ('books', 141), ('inspiration', 87), ('reading', 86), ('happiness', 68), ('writing', 67), ('inspirational-quotes', 65), ('friendship', 63), ('wisdom', 62), ('truth', 61), ('philosophy', 60), ('optimism', 48), ('living', 48), ('life-quotes', 48), ('death', 47), ('hope', 46), ('poetry', 45), ('romance', 45), ('positive-affirmation', 41), ('positive-thinking', 41), ('women', 40), ('life-lessons', 40), ('optimistic', 40), ('positive', 40), ('positive-life', 40), ('inspire', 38), ('inspiring', 38), ('courage', 37), ('life-and-living', 36), ('change', 34), ('pain', 34), ('funny', 34), ('inspirational-life', 34), ('relationships', 33), ('religion', 32), ('dreams', 31), ('music', 31), ('fear', 31), ('jace-wayland', 29), ('friends', 28), ('strength', 28), ('freedom', 28), ('god', 28), ('inspirational-attitude', 28), ('motivational', 27), ('loneliness', 27), ('science', 26), ('beauty', 26), ('time', 25), ('motivation', 25), (

In [18]:
# printing the top 40 most frequent tags
for i in range(0, len(tag_freq_sorted)):
    print(tag_freq_sorted[i])
    if i == 40:
        break


('love', 327)
('inspirational', 319)
('life', 295)
('humor', 254)
('books', 141)
('inspiration', 87)
('reading', 86)
('happiness', 68)
('writing', 67)
('inspirational-quotes', 65)
('friendship', 63)
('wisdom', 62)
('truth', 61)
('philosophy', 60)
('optimism', 48)
('living', 48)
('life-quotes', 48)
('death', 47)
('hope', 46)
('poetry', 45)
('romance', 45)
('positive-affirmation', 41)
('positive-thinking', 41)
('women', 40)
('life-lessons', 40)
('optimistic', 40)
('positive', 40)
('positive-life', 40)
('inspire', 38)
('inspiring', 38)
('courage', 37)
('life-and-living', 36)
('change', 34)
('pain', 34)
('funny', 34)
('inspirational-life', 34)
('relationships', 33)
('religion', 32)
('dreams', 31)
('music', 31)
('fear', 31)


From the above we can see two main issues with the tags as highlighted in blue and yellow:
1. We can see different columns which conceptually mean the same thing however it is written in different ways. For example, inspirational, inspiration, inspire and inspiring – all mean the same thing. However, it is classified as different tags. For fixing this issue, we must perform lemmatization on the target column.
2. We can see multiple different words separated by a hyphen. For example, positive-affirmation, life-lessons, etc. The hyphenated words must be handled by removing the ‘-‘ symbol. For example, ‘positive-thinking’ will be replaced with the tags positivethinking.

# Data Cleaning
## Seperating Words That Contain a Hyphen

In [19]:
dataset['tags'][0]

['be-yourself',
 'gilbert-perreira',
 'honesty',
 'inspirational',
 'misattributed-oscar-wilde',
 'quote-investigator']

In [20]:
type(dataset['tags'][0][0])

str

In [21]:
# removing '-' from hyphenated words:
for i in range(0, len(dataset)):
    for j in range(0, len(dataset['tags'][i])):
        # in this loop we can go through individual tag elements
        if('-' in dataset['tags'][i][j]):
            # here the word contains a '-' character, therefore we can split the data
            # print(dataset['tags'][i][j])
            # new_split_strings = str(dataset['tags'][i][j]).split('-')
            dataset['tags'][i][j] = dataset['tags'][i][j].replace("-", "")
            # print(dataset['tags'][i])
            # print("This is the string: "),
            # print(new_split_strings)
        # print(dataset['tags'][i][j]),
    # print(dataset['tags'][i])
    # below code was written for testing and as a safety measure
    # if (i == 10):
    #     break
# print((dataset['tags'][0:11]))

In [22]:
dataset['tags']

0       [beyourself, gilbertperreira, honesty, inspira...
1       [best, life, love, mistakes, outofcontrol, tru...
2       [humannature, humor, infinity, philosophy, sci...
3                                          [books, humor]
4                                   [books, simile, soul]
                              ...                        
2503                               [morality, philosophy]
2504                                 [happiness, success]
2505                                  [inspirationallife]
2506                              [comfort, home, winter]
2507        [justlisten, loud, owen, saradessen, silence]
Name: tags, Length: 2508, dtype: object

In [23]:
# writing a function to calculate the count of tags and returning a list of tuples - [(tag, count of tag)]
def calculate_count_of_tags(tags=dataset['tags']):
    tag_count_dict = {}
    for row in tags:
        for tag in row:
            if tag in tag_count_dict:
                tag_count_dict[tag] += 1
            else:
                tag_count_dict[tag] = 1
    # now sorting it based on the highest values
    tag_freq_sorted = sorted(tag_count_dict.items(), key=lambda x:x[1], reverse=True)
    return tag_freq_sorted
    

In [24]:
tag_freq_sorted = calculate_count_of_tags(tags=dataset['tags'])

In [30]:
# looking at the top 40 most frequent tags:
for i in range(0, len(tag_freq_sorted)):
    print(tag_freq_sorted[i])
    if(i==39):
        break

('love', 327)
('inspirational', 319)
('life', 295)
('humor', 254)
('books', 141)
('inspiration', 87)
('reading', 86)
('happiness', 68)
('writing', 67)
('inspirationalquotes', 65)
('friendship', 63)
('wisdom', 62)
('truth', 61)
('philosophy', 60)
('optimism', 48)
('living', 48)
('lifequotes', 48)
('death', 47)
('hope', 46)
('poetry', 45)
('romance', 45)
('positiveaffirmation', 41)
('positivethinking', 41)
('women', 40)
('lifelessons', 40)
('optimistic', 40)
('positive', 40)
('positivelife', 40)
('inspire', 38)
('inspiring', 38)
('courage', 37)
('lifeandliving', 36)
('change', 34)
('pain', 34)
('funny', 34)
('inspirationallife', 34)
('relationships', 33)
('religion', 32)
('dreams', 31)
('music', 31)


## Performing Lemmetization on Target Column

In [35]:
import nltk
from nltk.stem import WordNetLemmatizer

nltk.download('wordnet')
lemmatizer = WordNetLemmatizer()

# lemmetizing all tags
for i in range(0, len(dataset)):
    for j in range(0, len(dataset['tags'][i])):
        lemma = lemmatizer.lemmatize(dataset['tags'][i][j])
        dataset['tags'][i][j] = lemma
        # print(f"Actual word {dataset['tags'][i][j]}. Lemmatized word by default: {lemma}")
    # if(i==30):
    #     break

[nltk_data] Downloading package wordnet to
[nltk_data]     /Users/rishabhkaushick/nltk_data...
[nltk_data]   Package wordnet is already up-to-date!


In [36]:
dataset['tags']

0       [beyourself, gilbertperreira, honesty, inspira...
1       [best, life, love, mistake, outofcontrol, trut...
2       [humannature, humor, infinity, philosophy, sci...
3                                           [book, humor]
4                                    [book, simile, soul]
                              ...                        
2503                               [morality, philosophy]
2504                                 [happiness, success]
2505                                  [inspirationallife]
2506                              [comfort, home, winter]
2507        [justlisten, loud, owen, saradessen, silence]
Name: tags, Length: 2508, dtype: object

In [37]:
# running the count of tags again
tag_freq_sorted = calculate_count_of_tags()
for i in range(0, len(tag_freq_sorted)):
    print(tag_freq_sorted[i])
    if(i==39):
        break

('love', 327)
('inspirational', 319)
('life', 295)
('humor', 254)
('book', 144)
('inspiration', 87)
('reading', 86)
('happiness', 68)
('writing', 67)
('inspirationalquotes', 65)
('friendship', 63)
('wisdom', 62)
('truth', 61)
('philosophy', 60)
('optimism', 48)
('living', 48)
('lifequotes', 48)
('death', 47)
('hope', 46)
('poetry', 45)
('romance', 45)
('woman', 42)
('positiveaffirmation', 41)
('positivethinking', 41)
('lifelessons', 40)
('optimistic', 40)
('positive', 40)
('positivelife', 40)
('dream', 38)
('inspire', 38)
('inspiring', 38)
('courage', 37)
('relationship', 36)
('lifeandliving', 36)
('change', 35)
('pain', 34)
('funny', 34)
('inspirationallife', 34)
('religion', 32)
('fear', 32)


In [38]:
test_dataset = dataset
# lemmetizing all tags to a noun
for i in range(0, len(test_dataset)):
    for j in range(0, len(test_dataset['tags'][i])):
        lemma = lemmatizer.lemmatize(test_dataset['tags'][i][j], pos='n')
        test_dataset['tags'][i][j] = lemma
        # print(f"Actual word {test_dataset['tags'][i][j]}. Lemmatized word by default: {lemma}")
    # if(i==30):
    #     break

# running the count of tags again
tag_freq_sorted = calculate_count_of_tags(test_dataset['tags'])
for i in range(0, len(tag_freq_sorted)):
    print(tag_freq_sorted[i])
    if(i==39):
        break

('love', 327)
('inspirational', 319)
('life', 295)
('humor', 254)
('book', 144)
('inspiration', 87)
('reading', 86)
('happiness', 68)
('writing', 67)
('inspirationalquotes', 65)
('friendship', 63)
('wisdom', 62)
('truth', 61)
('philosophy', 60)
('optimism', 48)
('living', 48)
('lifequotes', 48)
('death', 47)
('hope', 46)
('poetry', 45)
('romance', 45)
('woman', 42)
('positiveaffirmation', 41)
('positivethinking', 41)
('lifelessons', 40)
('optimistic', 40)
('positive', 40)
('positivelife', 40)
('dream', 38)
('inspire', 38)
('inspiring', 38)
('courage', 37)
('relationship', 36)
('lifeandliving', 36)
('change', 35)
('pain', 34)
('funny', 34)
('inspirationallife', 34)
('religion', 32)
('fear', 32)


However, we still can make it better. We can see some values which are similar like inspirational, inspiration, inspiring, and inspire. We can use the concept of stemming to further improve our target attribute - tags.

## Stemming on Target Column ('Tags')
### Selection of Stemming Algorithm

I decided to use and compare between Snowball and Lancaster Stemmer in this project for the following reasons:
- Snowball Stemmer is an improved version of Porter Stemmer.
- The Lancaster Stemmer has a higher chance of over-stemming (this phrase refers to those times when two different words lead to the same stem) due to its aggressive algorithm selection method in an iterative manner.

#### Snowball Stemmer

In [47]:
from nltk.stem import SnowballStemmer

snowball_stemmer = SnowballStemmer("english")
# stem3 = stemmer3.stem(word)
# print(f"\nStemmed word using SnowballStemmer: {stem3}")
print("Using Snowball Stemmer: ")
for i in range(0, len(test_dataset)):
    for j in range(0, len(test_dataset['tags'][i])):
        stem = snowball_stemmer.stem(test_dataset['tags'][i][j])
        # test_dataset['tags'][i][j] = stem
        print(f"Actual word {test_dataset['tags'][i][j]}. Stemmed word : {stem}")
    if(i==30):
        break

Using Snowball Stemmer: 
Actual word beyourself. Stemmed word : beyourself
Actual word gilbertperreira. Stemmed word : gilbertperreira
Actual word honesty. Stemmed word : honesti
Actual word inspirational. Stemmed word : inspir
Actual word misattributedoscarwilde. Stemmed word : misattributedoscarwild
Actual word quoteinvestigator. Stemmed word : quoteinvestig
Actual word best. Stemmed word : best
Actual word life. Stemmed word : life
Actual word love. Stemmed word : love
Actual word mistake. Stemmed word : mistak
Actual word outofcontrol. Stemmed word : outofcontrol
Actual word truth. Stemmed word : truth
Actual word worst. Stemmed word : worst
Actual word humannature. Stemmed word : humannatur
Actual word humor. Stemmed word : humor
Actual word infinity. Stemmed word : infin
Actual word philosophy. Stemmed word : philosophi
Actual word science. Stemmed word : scienc
Actual word stupidity. Stemmed word : stupid
Actual word universe. Stemmed word : univers
Actual word book. Stemmed wor

#### Lancaster Stemmer

In [48]:
from nltk.stem import LancasterStemmer

lancaster_stemmer = LancasterStemmer()
print("Using Lancaster Stemmer")
for i in range(0, len(test_dataset)):
    for j in range(0, len(test_dataset['tags'][i])):
        stem = lancaster_stemmer.stem(test_dataset['tags'][i][j])
        # test_dataset['tags'][i][j] = stem
        print(f"Actual word {test_dataset['tags'][i][j]}. Stemmed word : {stem}")
    if(i==30):
        break

Using Lancaster Stemmer
Actual word beyourself. Stemmed word : beyourself
Actual word gilbertperreira. Stemmed word : gilbertperreir
Actual word honesty. Stemmed word : honesty
Actual word inspirational. Stemmed word : inspir
Actual word misattributedoscarwilde. Stemmed word : misattributedoscarwild
Actual word quoteinvestigator. Stemmed word : quoteinvestig
Actual word best. Stemmed word : best
Actual word life. Stemmed word : lif
Actual word love. Stemmed word : lov
Actual word mistake. Stemmed word : mistak
Actual word outofcontrol. Stemmed word : outofcontrol
Actual word truth. Stemmed word : tru
Actual word worst. Stemmed word : worst
Actual word humannature. Stemmed word : human
Actual word humor. Stemmed word : hum
Actual word infinity. Stemmed word : infin
Actual word philosophy. Stemmed word : philosoph
Actual word science. Stemmed word : sci
Actual word stupidity. Stemmed word : stupid
Actual word universe. Stemmed word : univers
Actual word book. Stemmed word : book
Actual w

#### Porter Stemmer

In [49]:
from nltk.stem import PorterStemmer

porter_stemmer = PorterStemmer()
print("Using Porter Stemmer")
for i in range(0, len(test_dataset)):
    for j in range(0, len(test_dataset['tags'][i])):
        stem = porter_stemmer.stem(test_dataset['tags'][i][j])
        # test_dataset['tags'][i][j] = stem
        print(f"Actual word {test_dataset['tags'][i][j]}. Stemmed word : {stem}")
    if(i==30):
        break

Using Porter Stemmer
Actual word beyourself. Stemmed word : beyourself
Actual word gilbertperreira. Stemmed word : gilbertperreira
Actual word honesty. Stemmed word : honesti
Actual word inspirational. Stemmed word : inspir
Actual word misattributedoscarwilde. Stemmed word : misattributedoscarwild
Actual word quoteinvestigator. Stemmed word : quoteinvestig
Actual word best. Stemmed word : best
Actual word life. Stemmed word : life
Actual word love. Stemmed word : love
Actual word mistake. Stemmed word : mistak
Actual word outofcontrol. Stemmed word : outofcontrol
Actual word truth. Stemmed word : truth
Actual word worst. Stemmed word : worst
Actual word humannature. Stemmed word : humannatur
Actual word humor. Stemmed word : humor
Actual word infinity. Stemmed word : infin
Actual word philosophy. Stemmed word : philosophi
Actual word science. Stemmed word : scienc
Actual word stupidity. Stemmed word : stupid
Actual word universe. Stemmed word : univers
Actual word book. Stemmed word : 

All 3 stemming approaches do not give out meaningful results in this case. Since our tags must be linguistically meaningful, we will not be performing stemming. 

## Removal of Stop Words in the Target Column ('Tags')
From the above, we can see that there are tags like 'of', 'the', 'and', 'to' which do not hold much meaning. These words are known as stop words. Therefore, I will perform the removal of stop words in the target column.

In [24]:
import nltk
from nltk.corpus import stopwords

nltk.download('stopwords')
print(stopwords.words('english'))

['i', 'me', 'my', 'myself', 'we', 'our', 'ours', 'ourselves', 'you', "you're", "you've", "you'll", "you'd", 'your', 'yours', 'yourself', 'yourselves', 'he', 'him', 'his', 'himself', 'she', "she's", 'her', 'hers', 'herself', 'it', "it's", 'its', 'itself', 'they', 'them', 'their', 'theirs', 'themselves', 'what', 'which', 'who', 'whom', 'this', 'that', "that'll", 'these', 'those', 'am', 'is', 'are', 'was', 'were', 'be', 'been', 'being', 'have', 'has', 'had', 'having', 'do', 'does', 'did', 'doing', 'a', 'an', 'the', 'and', 'but', 'if', 'or', 'because', 'as', 'until', 'while', 'of', 'at', 'by', 'for', 'with', 'about', 'against', 'between', 'into', 'through', 'during', 'before', 'after', 'above', 'below', 'to', 'from', 'up', 'down', 'in', 'out', 'on', 'off', 'over', 'under', 'again', 'further', 'then', 'once', 'here', 'there', 'when', 'where', 'why', 'how', 'all', 'any', 'both', 'each', 'few', 'more', 'most', 'other', 'some', 'such', 'no', 'nor', 'not', 'only', 'own', 'same', 'so', 'than', '

[nltk_data] Downloading package stopwords to
[nltk_data]     /Users/rishabhkaushick/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


In [25]:
# creating a set of stop words
stop_words_set = set(stopwords.words('english'))

for i in range(0, len(dataset)):
    for j in range(0, len(dataset['tags'][i])):
        print(dataset['tags'][i])
    if(i==10):
        break
        # if(dataset['tags'][i][j] in stop_words_set):
        #     dataset['tags'][i].pop(j)

['be-yourself', 'gilbert-perreira', 'honesty', 'inspirational', 'misattributed-oscar-wilde', 'quote-investigator']
['be-yourself', 'gilbert-perreira', 'honesty', 'inspirational', 'misattributed-oscar-wilde', 'quote-investigator']
['be-yourself', 'gilbert-perreira', 'honesty', 'inspirational', 'misattributed-oscar-wilde', 'quote-investigator']
['be-yourself', 'gilbert-perreira', 'honesty', 'inspirational', 'misattributed-oscar-wilde', 'quote-investigator']
['be-yourself', 'gilbert-perreira', 'honesty', 'inspirational', 'misattributed-oscar-wilde', 'quote-investigator']
['be-yourself', 'gilbert-perreira', 'honesty', 'inspirational', 'misattributed-oscar-wilde', 'quote-investigator']
['best', 'life', 'love', 'mistakes', 'truth', 'worst', 'out', 'of', 'control']
['best', 'life', 'love', 'mistakes', 'truth', 'worst', 'out', 'of', 'control']
['best', 'life', 'love', 'mistakes', 'truth', 'worst', 'out', 'of', 'control']
['best', 'life', 'love', 'mistakes', 'truth', 'worst', 'out', 'of', 'cont

In [31]:
dataset['tags'][0]

['be-yourself',
 'gilbert-perreira',
 'honesty',
 'inspirational',
 'misattributed-oscar-wilde',
 'quote-investigator']

In [34]:
dataset

quote  \
0        “Be yourself; everyone else is already taken.”   
1     “I'm selfish, impatient and a little insecure....   
2     “Two things are infinite: the universe and hum...   
3                      “So many books, so little time.”   
4     “A room without books is like a body without a...   
...                                                 ...   
2503  “Morality is simply the attitude we adopt towa...   
2504  “Don't aim at success. The more you aim at it ...   
2505  “In life, finding a voice is speaking and livi...   
2506  “Winter is the time for comfort, for good food...   
2507                      “Silence is so freaking loud”   

                     author                                               tags  
0               Oscar Wilde  [be-yourself, gilbert-perreira, honesty, inspi...  
1            Marilyn Monroe  [best, life, love, mistakes, truth, worst, con...  
2           Albert Einstein  [humor, infinity, philosophy, science, stupidi...  
3               Frank Zappa                                     [books, humor]  
4     Marcus Tullius Cicero                              [books, simile, soul]  
...                     ...                                                ...  
2503           Oscar Wilde,                             [morality, philosophy]  
2504      Viktor E. Frankl,                               [happiness, success]  
2505           John Grisham                              [inspirational, life]  
2506          Edith Sitwell                            [comfort, home, winter]  
2507          Sarah Dessen,  [loud, owen, silence, just, listen, sara, dessen]  

[2508 rows x 3 columns]

# Multi-Label Naive Bayes Classifier (Baseline Model)

# Recurrant Neural Network (RNN) Model

# Transformer Based Model?